cmd/dp-import-tracker/main.go

```diff
@@ -107,45 +115,83 @@ func CheckImportJobCompletionState(importAPI *api.ImportAPI, datasetAPI *api.Dat
 			continue
 		}
 		// XXX TODO code below largely untested, and possibly subject to race conditions
-		instanceFromAPI, err := datasetAPI.GetInstance(instanceRef.ID)
+		instanceFromAPI, err, isFatal := datasetAPI.GetInstance(ctx, instanceRef.ID)
 		if err != nil {
-			return err
+			return err, isFatal
 		}
 		if instanceFromAPI.State != "completed" && instanceFromAPI.State != "error" {
-			return nil
+			return nil, false
 		}
 		if instanceFromAPI.State == "error" {
 			targetState = instanceFromAPI.State
 		}
 	}
 	// assert: all instances for jobID are marked "completed"/"error", so update import as same
-	if err := importAPI.UpdateImportJobState(jobID, targetState); err != nil {
+	if err := importAPI.UpdateImportJobState(ctx, jobID, targetState); err != nil {
 		log.ErrorC("CheckImportJobCompletionState update", err, log.Data{"jobID": jobID, "last completed instanceID": completedInstanceID})
 	}
-	return nil
+	return nil, false
 }
 
 // manageActiveInstanceEvents handles all updates to trackedInstances in one thread (this is only called once, in its own thread)
 func manageActiveInstanceEvents(
+	ctx context.Context,
 	createInstanceChan chan string,
 	updateInstanceWithObservationsInsertedChan chan insertedObservationsEvent,
 	datasetAPI *api.DatasetAPI,
-	importAPI *api.ImportAPI) {
+	importAPI *api.ImportAPI,
+	instanceLoopDoneChan chan bool) {
+
+	// inform main() when we have stopped processing events
+	defer close(instanceLoopDoneChan)
 
 	trackedInstances := make(trackedInstanceList)
-	if err := trackedInstances.getInstanceList(datasetAPI); err != nil {
+	if err, _ := trackedInstances.getInstanceList(ctx, datasetAPI); err != nil {
 		logFatal("could not obtain initial instance list", err, nil)
 	}
 
-	checkForCompletedInstancesChan := make(chan bool)
+	tickerChan := make(chan bool)
 	go func() {
 		for range time.Tick(checkForCompleteInstancesTick) {
-			checkForCompletedInstancesChan <- true
+			tickerChan <- true
 		}
 	}()
 
-	for {
+	for looping := true; looping; {
```

Comment:
If this is just to deal with breaking out of a nested select, naming the loop might be more idiomatic: https://www.goinggo.net/2013/11/label-breaks-in-go.html

Reply:
I originally had a `break` but I actually prefer this.